### PR TITLE
Implement Tail Call Optimize into ReturnCall Instruction

### PR DESCRIPTION
**Pull request by makachanm**
I implemented a tail call optimize into WebAssembly 3's ReturnCall recursive instruction.
It currently unfinished, (the issue with type checking in the standard's spec, indirect call is not completed..) but it passes [Return Call Test](https://github.com/Samsung/walrus/blob/main/test/web-assembly3/gc/return_call.wast) code.

### CODE REVIEW

**Comment by clover2123** [comment]
Please run `tools/check_tidy.py --update` to format the code according to the style guidelines.

**Comment by makachanm** [comment]
I applied. thanks.

File: src/interpreter/Interpreter.h [hunk 1]
```diff
@@ -24,6 +24,7 @@
 #include "runtime/Module.h"
 #include "runtime/Tag.h"
 #include "interpreter/ByteCode.h"
+#include <utility>
```
Comment (zherczeg) [inline src/interpreter/Interpreter.h]: Why do you need this?

File: src/runtime/ExecutionState.h [hunk 2]
```diff
@@ -19,6 +19,8 @@
 
 #include "util/Optional.h"
 #include "util/Util.h"
+#include <cstddef>
+#include <vector>
```
Comment (zherczeg) [inline src/runtime/ExecutionState.h]: Do you need these?

File: src/runtime/ExecutionState.h [hunk 3]
```diff
@@ -53,6 +59,22 @@ class ExecutionState {
...
+        tco_functionTarget = nullptr;
+    }
+
+    bool hasTCO()
+    {
+        return tco_functionTarget != nullptr;
+    }
+
+    std::vector<size_t> tco_paramStore;
+    size_t tco_paramSize;
```
Comment (zherczeg) [inline src/runtime/ExecutionState.h]: Vector should have a size.

File: src/interpreter/ByteCode.h [hunk 4]
```diff
@@ -2216,6 +2217,59 @@ class CallRef : public ByteCode {
...
+    {
+    }
+
+    uint32_t index() const { return m_index; }
+    ByteCodeStackOffset* stackOffsets() const
+    {
+        return reinterpret_cast<ByteCodeStackOffset*>(reinterpret_cast<size_t>(this) + sizeof(Call));
+    }
+
+    uint16_t parameterOffsetsSize() const
```
Comment (zherczeg) [inline src/interpreter/ByteCode.h:2151]: Correct me if I wrong, but as far as I remember parameters are always stored in a linear memory area. This means their offsets do not need to be stored, they can be computed during running. The size of the tco buffer could be stored so the buffer could be pre-allocated.

**Review by zherczeg — COMMENTED** [review]
This sounds like a good idea. I would simplify the parameter storing: the parameters are simply stored in a linear buffer, and copied by a memcpy.

**Comment by zherczeg** [comment]
Parameters are stored in a linear area:
https://github.com/Samsung/walrus/blob/main/src/interpreter/Interpreter.h#L57

Then the tco buffer could be copied here without changes.

**Comment by makachanm** [comment]
I'll rework parameter recover code with  memcpy and Walrus's own vector implementation. Thanks for the feedback.

File: src/runtime/ExecutionState.h (hunk 3 above)
Comment (clover2123) [reply src/runtime/ExecutionState.h]: I recommend using the predefined Vector structure as shown below.
Also, regarding naming conventions, the Walrus project typically uses member variables that start with `m_` and avoids underscores in the middle of names.
```c++
Vector<size_t> m_tcoParamStore;
```

If you are using a fixed-length vector, please use VectorWithFixedSize instead:
```c++
VectorWithFixedSize<size_t> m_tcoParamStore;
m_tcoParamStore.reserve(newSize);  // allocate array
m_tcoParamStore.clear();           // reset array
```

**Comment by makachanm** [comment]
I think lack of JIT version of TCO implement causes CI failure. I'll seperate it for future works.

File: src/runtime/ExecutionState.h [hunk 5]
```diff
@@ -53,6 +56,20 @@ class ExecutionState {
         return m_stackLimit;
     }
 
+    void destroyTCO()
+    {
+        m_tcoparamStore.clear();
+        m_tcofunctionTarget = nullptr;
```
Comment (zherczeg) [inline src/runtime/ExecutionState.h]: The names are wrong. They should be m_tcoParamStore and m_tcoFunction store.

**Comment by makachanm** [comment]
I've verified that all tests in the walrus suite are passing in the x64 local machine at interpret mode.

**Comment by clover2123** [comment]
There is a conflict in `ByteCode.h` file. 
Please rebase it and push this pr again

**Comment by makachanm** [comment]
CI is failing when Clang compiler is used. I had no problem with GCC.
I'll fix it with reordering some parameter initialize order.

File: src/interpreter/Interpreter.cpp [hunk 6]
```diff
@@ -1711,6 +1712,75 @@ ByteCodeStackOffset* Interpreter::interpret(ExecutionState& state,
...
+    }
+
+    DEFINE_OPCODE(ReturnCallIndirect)
+        :
+    {
+        ReturnCallIndirect* code = (ReturnCallIndirect*)programCounter;
+        Table* table = instance->table(code->tableIndex());
+
+        uint32_t idx = readValue<uint32_t>(bp, code->calleeOffset());
+        if (idx >= table->size()) {
```
Comment (clover2123) [inline src/interpreter/Interpreter.cpp:1738]: Add `UNLIKELY` to this check code

File: src/interpreter/Interpreter.cpp [hunk 7]
```diff
@@ -1711,6 +1712,75 @@ ByteCodeStackOffset* Interpreter::interpret(ExecutionState& state,
...
+            Trap::throwException(state, "undefined element");
+        }
+        auto target = reinterpret_cast<Function*>(table->uncheckedGetElement(idx));
+        if (UNLIKELY(Value::isNull(target))) {
+            Trap::throwException(state, "uninitialized element " + std::to_string(idx));
+        }
+        const FunctionType* ft = target->functionType();
+        if (!ft->equals(code->functionType())) {
```
Comment (clover2123) [inline src/interpreter/Interpreter.cpp:1746]: `UNLIKELY` here too

File: src/interpreter/Interpreter.cpp [hunk 8]
```diff
@@ -1711,6 +1712,75 @@ ByteCodeStackOffset* Interpreter::interpret(ExecutionState& state,
...
+        :
+    {
+        ReturnCallRef* code = (ReturnCallRef*)programCounter;
+
+        auto target = readValue<Function*>(bp, code->calleeOffset());
+        if (UNLIKELY(Value::isNull(target))) {
+            Trap::throwException(state, "null function reference");
+        }
+        const FunctionType* ft = target->functionType();
+        if (!ft->equals(code->functionType())) {
```
Comment (clover2123) [inline src/interpreter/Interpreter.cpp:1770]: ditto

File: src/interpreter/Interpreter.h [hunk 9]
```diff
@@ -106,6 +110,13 @@ class Interpreter {
             resultOffsets = interpret(newState, programCounter, functionStackBase, function->instance());
         }
 
+        if (newState.hasTCO()) {
+            state.m_tcoParamStore.reserve(newState.m_tcoParamStore.size());
+            VectorCopier<size_t>::copy(state.m_tcoParamStore.data(), newState.m_tcoParamStore.data(), newState.m_tcoParamStore.size());
+            state.m_tcoFunctionTarget = newState.m_tcoFunctionTarget;
+            return;
+        }
```
Comment (clover2123) [inline src/interpreter/Interpreter.h:118]: At this point, the vector needs to be copied into the upper `ExecutionState`.
Instead of this copy, what about allocating one global vector and reusing this vector?
This approach could reduce copy overhead.

File: src/interpreter/Interpreter.h [hunk 10]
```diff
@@ -132,6 +143,11 @@ class Interpreter {
                                  uint8_t* bp,
                                  Instance* instance);
 
+    static void returnCallOperation(ExecutionState& state,
```
Comment (clover2123) [inline src/interpreter/Interpreter.h:146]: It seems that this function is declared but never used

File: src/interpreter/Interpreter.h [hunk 11]
```diff
@@ -52,9 +52,13 @@ class Interpreter {
         auto moduleFunction = function->moduleFunction();
         ALLOCA(uint8_t, functionStackBase, moduleFunction->requiredStackSize());
 
-        // init parameter space
-        for (size_t i = 0; i < parameterOffsetCount; i++) {
-            ((size_t*)functionStackBase)[i] = *((size_t*)(bp + offsets[i]));
+        if (state.hasTCO()) {
+            VectorCopier<size_t>::copy((size_t*)functionStackBase, state.m_tcoParamStore.data(), state.m_tcoParamStore.size());
+            state.destroyTCO();
```
Comment (clover2123) [inline src/interpreter/Interpreter.h:57]: When interpreter is called again due to tail call, 
`state.destroyTCO();` has been called so `state` doesn't have any valid TCO data at this point.
Would you check it again?

**Comment by clover2123** [comment]
Lastly, have you been able to test this patch for recursive tail calls?

File: src/interpreter/Interpreter.h (hunk 11 above)
Comment (makachanm) [reply src/interpreter/Interpreter.h:57]: I'll check it later.

File: src/interpreter/Interpreter.h (hunk 9 above)
Comment (makachanm) [reply src/interpreter/Interpreter.h:118]: I don't have a idea for where I declare a global TCO buffer. I think addition of new execution state manager will be affect the code's consistency. and It can be harder to decide when the timing the destroy a TCO buffer.

**Comment by makachanm** [comment]
> Lastly, have you been able to test this patch for recursive tail calls?

I'll do it through recursive fibonacci test. Thanks for the mention.

**Comment by makachanm** [comment]
I found fibonacci test makes call stack exhaustion. I'm analyzing what's problem with.

File: src/interpreter/Interpreter.h (hunk 9 above)
Comment (clover2123) [reply src/interpreter/Interpreter.h:118]: I think that `Store` would be a good place to keep the TCO buffer.
`Store` can be accessed through `instance->module()->store()` in the interpreter.

**Comment by makachanm** [comment]
Yup it passes 10000-depth fibonacci test without problems. 
I removed some edge-case code so it may cause problem in JIT, but it's seems not a huge problem.

I'll add new edge-case logic into it.

File: src/interpreter/Interpreter.h [hunk 12]
```diff
@@ -106,6 +112,10 @@ class Interpreter {
             resultOffsets = interpret(newState, programCounter, functionStackBase, function->instance());
         }
 
+        if (store->hasTCO()) {
+            return;
```
Comment (clover2123) [inline src/interpreter/Interpreter.h]: What about adding an assertion to guarantee that TCO  has triggered, like below?
```
ASSERT(!resultOffsets); // resultOffsets should be nullptr
```

File: src/parser/WASMParser.cpp [hunk 13]
```diff
@@ -1473,6 +1475,11 @@ class WASMBinaryReader : public wabt::WASMBinaryReaderDelegate {
             offsetIndex += subIndexCount;
         }
 
+        if (isReturnCall) {
+            ASSERT(offsetIndex == code->parameterOffsetsSize());
```
Comment (clover2123) [inline src/parser/WASMParser.cpp]: Is it fine to return immediately here for tail call?
In this case, result offsets are not calculated.

File: src/interpreter/Interpreter.h [hunk 14]
```diff
@@ -50,11 +51,16 @@ class Interpreter {
...
         auto moduleFunction = function->moduleFunction();
+        auto store = function->instance()->module()->store();
         ALLOCA(uint8_t, functionStackBase, moduleFunction->requiredStackSize());
 
-        // init parameter space
-        for (size_t i = 0; i < parameterOffsetCount; i++) {
-            ((size_t*)functionStackBase)[i] = *((size_t*)(bp + offsets[i]));
+        if (store->hasTCO()) {
+            VectorCopier<size_t>::copy((size_t*)functionStackBase, store->tcoBuffer(), store->tcoBufferSize());
+            store->clearTCO();
```
Comment (clover2123) [inline src/interpreter/Interpreter.h]: In tail call optimization, only the parameters are copied and delivered just before the interpreter's execution.
But IMO we need to copy the result offsets too, because they could differ if the tail call isn't a recursive call.

**Comment by clover2123** [comment]
One more, this patch doesn't consider exceptions.
What if exception occurs during the execution of tail call?

File: src/parser/WASMParser.cpp (hunk 13 above)
Comment (makachanm) [reply src/parser/WASMParser.cpp]: In the trampoline loop, we need to mark where we start execution again. This code is used to mark the place where the interpreter is begin execution again. We just needed to mark how many parameters exists on procedure.

**Comment by makachanm** [comment]
I found some inconsistency in parameter space. Copy to callee can be not a problem, but Copy from caller is making some corruption into buffer value. 

I changed it does manually copy data one by one.

File: src/parser/WASMParser.cpp (hunk 13 above)
Comment (makachanm) [reply src/parser/WASMParser.cpp]: but I'll remove that code. It seems to keep old code will not be a serious performance disadvantage.

File: src/interpreter/Interpreter.h (hunk 12 above)
Comment (makachanm) [reply src/interpreter/Interpreter.h]: Oops that makes segment fault. I'll keep current version.

File: src/interpreter/Interpreter.h (hunk 14 above)
Comment (makachanm) [reply src/interpreter/Interpreter.h]: I'll do it right now.

**Comment by makachanm** [comment]
Okay I rewrite all trampoline loop code into in-function flow control TCO. 
I checked the all built-in suite tests are passing, and I need to final check in CI.

File: test/wasi/write_to_this.txt [hunk 15]
```diff
@@ -1 +0,0 @@
-Hello World!
```
Comment (zherczeg) [inline test/wasi/write_to_this.txt:1]: Is this file needs to be deleted?

File: .gitignore [hunk 16]
```diff
@@ -19,3 +19,10 @@ rules.ninja
 node_modules
 /test/wasmBenchmarker/ctests/wasm
 /test/wasmBenchmarker/emsdk
+/.cache
+/build/.cache
+/build/.cmake
+/build/_deps
+/build/third_party
+/build/walrus
+compile_commands.json
```
Comment (zherczeg) [inline .gitignore]: What are these files?

File: src/parser/WASMParser.cpp [hunk 17]
```diff
@@ -1593,6 +1595,71 @@ class WASMBinaryReader : public wabt::WASMBinaryReaderDelegate {
         generateCallExpr(code, parameterCount, resultCount, functionType);
     }
 
+    virtual void OnReturnCallExpr(uint32_t index) override
+    {
+        m_preprocessData.seenBranch();
+        if (UNLIKELY(m_useJIT)) {
```
Comment (zherczeg) [inline src/parser/WASMParser.cpp]: Why JIT use is unlikely?

File: src/interpreter/Interpreter.cpp [hunk 18]
```diff
@@ -3091,10 +3165,56 @@ NEVER_INLINE void Interpreter::callRefOperation(
...
+    Instance*& instance,
+    Function* target,
+    ByteCodeStackOffset* offsets,
+    uint16_t parameterOffsetCount,
+    uint16_t resultOffsetCount)
+{
+    if (LIKELY(target->kind() == Function::DefinedFunctionKind)) {
+        DefinedFunction* definedTarget = target->asDefinedFunction();
+        ModuleFunction* targetModuleFunction = definedTarget->moduleFunction();
+        bool isInterpretableInPlace = frame.hasTryCatchContext() || !targetModuleFunction->hasTryCatch();
```
Comment (zherczeg) [inline src/interpreter/Interpreter.cpp]: Since walrus uses a zero cost try/catch model, the `considerException` in the `interpreter.h` might be not necessary. It would be good to test if it really has any advantage, and then just delete it, which could make this check unnecessary.

File: test/wasi/write_to_this.txt (hunk 15 above)
Comment (makachanm) [reply test/wasi/write_to_this.txt:1]: That's just placeholder for works. I forgot to delete it.

File: .gitignore (hunk 16 above)
Comment (makachanm) [reply .gitignore]: For my build situation. I'll delete it and reconfigure a cmake.

File: src/parser/WASMParser.cpp (hunk 17 above)
Comment (makachanm) [reply src/parser/WASMParser.cpp]: This works is currently only for the interpreter. JIT seems to not work with newer instruction.
so marking & different codegen is needed to working with JIT things.

Comment (zherczeg) [reply src/parser/WASMParser.cpp]: I understand this. The question was why the bool is unlikely? You expect that jit is rarely used?

Comment (makachanm) [reply src/parser/WASMParser.cpp]: I thought UNLIKELY was need to parser branch. I'll remove it.

Comment (kulcsaradam) [reply src/parser/WASMParser.cpp]: > I thought UNLIKELY was need to parser branch. I'll remove it.

UNLIKELY is a macro for gcc and clang for better branch prediction. It's defined in `Walrus.h`: https://github.com/Samsung/walrus/blob/96448e4b94f5ddf321cd34879608ef3577621d67/src/Walrus.h#L69

File: .gitignore [hunk 19]
```diff
@@ -19,3 +19,7 @@ rules.ninja
 node_modules
 /test/wasmBenchmarker/ctests/wasm
 /test/wasmBenchmarker/emsdk
+/.cache
+/build/.cache
```
Comment (zherczeg) [inline .gitignore]: Why don't you build in the `out/` directory?

File: test/wasi/write_to_this.txt (hunk 15 above)
Comment (zherczeg) [inline test/wasi/write_to_this.txt:1]: Still removed.

**Review by zherczeg — COMMENTED** [review]
I think this patch is close to ready

**Comment by zherczeg** [comment]
I think you overdid the removal. Please check the merged version of your patch. This patch is a feature patch, which should not change any build files or existing helper files in the project (unless this patch affects them). If something is really unnecessary, you can delete it later in another patch. To use the out: `ninja -v -Cout`

**Comment by makachanm** [comment]
Oh I realized that two file was for checking WASI functionality. I'll revert that. Sorry for hasty decisions.

File: .gitignore [hunk 20]
```diff
@@ -19,3 +19,7 @@ rules.ninja
 node_modules
 /test/wasmBenchmarker/ctests/wasm
 /test/wasmBenchmarker/emsdk
+/.cache
```
Comment (zherczeg) [inline .gitignore]: These lines are still added.

**Comment by zherczeg** [comment]
Please squash the patch or check the "Files changed" tab here, and restore any changes that is not related to the patch.

File: src/interpreter/Interpreter.h [hunk 21]
```diff
@@ -52,26 +106,30 @@ class Interpreter {
...
         size_t programCounter = reinterpret_cast<size_t>(moduleFunction->byteCode());
+        StackFrame frame(functionStackBase, moduleFunction->requiredStackSize());
         ByteCodeStackOffset* resultOffsets;
 
 #if defined(WALRUS_ENABLE_JIT)
         if (moduleFunction->jitFunction() != nullptr) {
             resultOffsets = moduleFunction->jitFunction()->call(newState, function->instance(), functionStackBase);
-        } else if (considerException) {
-#else
-        if (considerException) {
```
Comment (clover2123) [inline src/interpreter/Interpreter.h:68]: Originally, exception is separately handled in `callInterpreter`, but this patch just merged it.
I think that this separation is mainly for performance purpose, such that fast/light execution for non-exceptional function.
Would this patch may affect the overall performance negatively?
I also wonder your opinion (@ksh8281 @zherczeg)

**Comment by clover2123** [comment]
I want to ask about exception handling in tail calls.
What if an exception occurs during the execution of tail calls?
Is this exception caught and handled as well?

File: src/interpreter/Interpreter.h (hunk 21 above)
Comment (zherczeg) [reply src/interpreter/Interpreter.h:68]: This part should be measured, but I think there should be no negative effect since try-catch is zero cost (when no exception), and a `while(true)` is just a backward jump.

**Comment by makachanm** [comment]
> I want to ask about exception handling in tail calls. What if an exception occurs during the execution of tail calls? Is this exception caught and handled as well?

Yes. I checked about it, and its catches exception well.
When the exception is occurred, the executor of callee is still original function itself.
so we can still use function that wrapped in try/catch.

when it's falls into fallback mode, it's internally uses `Interpreter.h`'s call.
so that can be protected by try/catch, and it propagates exception into call stack.

**Review by clover2123 — APPROVED** [review]
LGTM

File: test/wasi/write_to_this.txt (hunk 15 above)
Comment (clover2123) [inline test/wasi/write_to_this.txt:1]: BTW, what is this file for?
Is this unnecessary?

Comment (makachanm) [reply test/wasi/write_to_this.txt:1]: It's need to test WASI's filesystem functions.

Comment (clover2123) [reply test/wasi/write_to_this.txt:1]: If so, it should not be removed
